clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -36,65 +36,85 @@
 
 syscall_pt_trace_t::syscall_pt_trace_t()
     : open_file_func_(nullptr)
     , write_file_func_(nullptr)
     , close_file_func_(nullptr)
     , pttracer_handle_ { GLOBAL_DCONTEXT, nullptr }
+    , pttracer_output_buffer_ { GLOBAL_DCONTEXT, nullptr }
     , recorded_syscall_count_(0)
     , cur_recording_sysnum_(-1)
     , drcontext_(nullptr)
-    , pt_dir_name_ { '\0' }
+    , output_file_(INVALID_FILE)
 {
 }
 
 syscall_pt_trace_t::~syscall_pt_trace_t()
 {
+    if (output_file_ != INVALID_FILE) {
+        close_file_func_(output_file_);
+        output_file_ = INVALID_FILE;
+    }
 }
 
 bool
-syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name, size_t pt_dir_name_size,
+syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name,
                          drmemtrace_open_file_func_t open_file_func,
                          drmemtrace_write_file_func_t write_file_func,
                          drmemtrace_close_file_func_t close_file_func)
 {
     drcontext_ = drcontext;
-    memcpy(pt_dir_name_, pt_dir_name, pt_dir_name_size);
     open_file_func_ = open_file_func;
     write_file_func_ = write_file_func;
     close_file_func_ = close_file_func;
     pttracer_handle_ = { drcontext, nullptr };
-    return true;
-}
-
-bool
-syscall_pt_trace_t::start_syscall_pt_trace(int sysnum)
-{
+    pttracer_output_buffer_ = { drcontext_, nullptr };
+    std::string output_file_name(pt_dir_name);
+    output_file_name +=
+        "/" + std::to_string(dr_get_thread_id(drcontext_)) + PT_DATA_FILE_NAME_SUFFIX;
+    output_file_ = open_file_func_(output_file_name.c_str(), DR_FILE_WRITE_REQUIRE_NEW);
 #define RING_BUFFER_SIZE_SHIFT 8
-    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
-    ASSERT(pttracer_handle_.handle == nullptr, "pttracer_handle_.handle isn't nullptr");
 
-    /* TODO i#5505: To reduce the overhead caused by pttracer initialization, we should
-     * share the same pttracer handle for all syscalls. We will do this when we upgrade
-     * drpt2ir to support decoding PT fragments that don't have PSB packets.
+    /* To reduce the overhead caused by pttracer initialization, we shared the same
+     * pttracer handle for all syscalls per thread.
      */
     if (drpttracer_create_handle(drcontext_, DRPTTRACER_TRACING_ONLY_KERNEL,
                                  RING_BUFFER_SIZE_SHIFT, RING_BUFFER_SIZE_SHIFT,
                                  &pttracer_handle_.handle) != DRPTTRACER_SUCCESS) {
         return false;
     }
+    if (drpttracer_create_output(drcontext_, RING_BUFFER_SIZE_SHIFT, 0,
+                                 &pttracer_output_buffer_.data) != DRPTTRACER_SUCCESS) {
+        return false;
+    }
+
+    /* All syscalls in the same thread share the same pttracer handle. So they shared the
+     * same pt_metadata.
+     */
+    pt_metadata_t pt_metadata;
+    if (drpttracer_get_pt_metadata(pttracer_handle_.handle, &pt_metadata)) {
+        return false;
+    }
+    if (!metadata_dump(pt_metadata)) {
```

Comment:
What happens if the user inits multiple times for the same thread? Probably this function will write the metadata multiple times, which is not good? If yes, document that this init routine should be called only once per thread and also add an `initialized` bool if you think it's needed.

Reply:
Done.